compiler/noirc_frontend/src/elaborator/types.rs

```diff
@@ -118,7 +118,15 @@ impl<'context> Elaborator<'context> {
                 let fields = self.resolve_type_inner(*fields, kind);
                 Type::FmtString(Box::new(resolved_size), Box::new(fields))
             }
-            Quoted(quoted) => Type::Quoted(quoted),
+            Quoted(quoted) => {
+                let in_function = matches!(self.current_item, Some(DependencyId::Function(_)));
+                if in_function && !self.in_comptime_context() {
```

Comment:
We only error when we're in a function to allow code like `type DeriveFunction = fn(FunctionDefinition) -> Quoted` or impls over comptime types.